### PR TITLE
Add firewall CLI and facades

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Add a new CLI for setting and listing firewall rules.
Most of this PR is api/apiserver boiler plate. There's also a driveby fix for the state firewall rule save api.
CMR will use the juju-application-offer rule to control allowed ingress for cross model relations.
Allowance has also been made for specifying rules ssh and controller api ingress (but those are not currently implemented).

## QA steps

bootstrap
$ juju set-firewall-rule ssh --blacklist 10.0.0.0/8 --whitelist 192.168.0.0/16
$ juju firewall-rules
Service        Whitelist subnets           Blacklist subnets
ssh              192.168.0.0/16              10.0.0.0/8

## Documentation changes

Doc will be needed when it's all done.

### CODE REVIEW

File: cmd/juju/firewall/listrules.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	client, err := c.newAPIFunc()
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+	rulesResult, err := client.ListFirewallRules()
+	if err != nil {
+		return err
+	}
+	if len(rulesResult) == 0 {
```
Comment (axw) [inline cmd/juju/firewall/listrules.go]: there's a bug saying that the output (for some other data type) should always be valid YAML/JSON if one of those is requested. can you please make them output an empty list?

File: state/firewallrules.go [hunk 2]
```diff
@@ -90,20 +90,21 @@ func NewFirewallRules(st *State) *firewallRulesState {
 }
 
 // Save stores the specified firewall rules.
-func (fw *firewallRulesState) Save(service WellKnownServiceType, whiteListCidrs, blackListCidrs []string) (*FirewallRule, error) {
-	if err := service.validate(); err != nil {
-		return nil, errors.Trace(err)
+func (fw *firewallRulesState) Save(rule *FirewallRule) error {
```
Comment (axw) [inline state/firewallrules.go]: please make it a non-pointer type. then you can be sure it's non-nil, and the caller can be sure that this method won't (can't) modify it

Comment (wallyworld) [reply state/firewallrules.go]: ok

File: cmd/juju/firewall/listrules.go (hunk 1 above)
Comment (wallyworld) [reply cmd/juju/firewall/listrules.go]: yep.

**Comment by wallyworld** [comment]
$$merge$$

**Comment by wallyworld** [comment]
$$merge$$

**Comment by nskaggs** [comment]
@pmatulis This will need docs. @wallyworld we'll also need some CI around this.

**Pull request merged**